tests/test_dragon_installer.py

```diff
@@ -0,0 +1,519 @@
+
+    # verify archive is gone after cleanup
+    assert not archive_path.exists()
+    assert not extract_path.exists()
+
+
+def test_cleanup_no_archive(test_dir: str) -> None:
+    """Ensure that the cleanup method doesn't bomb when called with
+    missing archive"""
+    archive_path = pathlib.Path(test_dir) / mock_archive_name
+    extract_path = pathlib.Path(test_dir) / mock_archive_name.replace(".tar.gz", "")
+
+    extract_path.mkdir()
+
+    # verify archive exists before cleanup
+    assert extract_path.exists()
+
+    cleanup(archive_path, extract_path)
+
+    # verify archive is gone after cleanup
+    assert not archive_path.exists()
+    assert not extract_path.exists()
+
+
+def test_expand_archive(test_dir: str) -> None:
+    """Verify archive is expanded into expected location w/correct content"""
+    archive_path = pathlib.Path(test_dir) / mock_archive_name
+    exp_path = pathlib.Path(test_dir) / mock_archive_name.replace(".tar.gz", "")
+    num_files = 10
+
+    # create an archive to clean up
+    with tarfile.TarFile.open(archive_path, mode="w:gz") as tar:
+        for i in range(num_files):
+            content = pathlib.Path(test_dir) / f"{i:04}.txt"
+            content.write_text(f"i am file {i}\n")
+            tar.add(content)
+
+    extract_path = expand_archive(archive_path)
+
+    files = list(extract_path.rglob("*.txt"))
+
+    assert len(files) == num_files
+    assert extract_path == exp_path
+
+
+def test_expand_archive_path_path(test_dir: str) -> None:
+    """Verify the expand method responds to a bad path with a ValueError"""
+    archive_path = pathlib.Path(test_dir) / mock_archive_name
+
+    with pytest.raises(ValueError) as ex:
+        expand_archive(archive_path)
+
+    assert str(archive_path) in str(ex.value.args[0])
+
+
+def test_retrieve_cached(test_dir: str) -> None:
+    """Verify that a previously retrieved asset archive is re-used"""
+    working_dir = pathlib.Path(test_dir)
+    archive_path = working_dir / mock_archive_name
+    num_files = 10
+
+    # create an archive to simulate re-use
+    with tarfile.TarFile.open(archive_path, mode="w:gz") as tar:
+        for i in range(num_files):
+            content = working_dir / f"{i:04}.txt"
+            content.write_text(f"i am file {i}\n")
+            tar.add(content)
+
+    ts1 = archive_path.stat().st_ctime
+
+    requester = Requester(
+        auth=None,
+        base_url="https://github.com",
+        user_agent="mozilla",
+        per_page=10,
+        verify=False,
+        timeout=1,
+        retry=1,
+        pool_size=1,
+    )
+    headers = {"mock-header": "mock-value"}
+    attributes = {"mock-attr": "mock-attr-value"}
+    completed = True
+
+    asset = GitReleaseAsset(requester, headers, attributes, completed)
+
+    # ensure mocked asset has values that we use...
+    setattr(asset, "_browser_download_url", _git_attr(value="http://foo"))
+    setattr(asset, "_name", _git_attr(value=mock_archive_name))
```

Comment:
Pedantic "consider using `monkeypatch`" comment here as well, but again shouldn't really matter as this isn't setting anything that will leak outside the test

Reply:
Done